### DISCUSSION TITLE
[A-L] (18 giugno 2024) Esercizio 1

### DISCUSSION THREAD

**Discussion by FeddyLix17** [sapienzastudentsnetwork/progettazione-di-algoritmi#91] [Esami A-L]
Dare lo pseudocodice di un algoritmo $\Large O(n^2)$ che prende in input un array $\Large A[]$ di interi distinti di lunghezza $\Large n$ e trova lunghezza massima di una sottoseguenza crescente possibile nell'array.

Per esempio, se $\Large A = [1,\ 4,\ 2,\ 6,\ 3,\ 8,\ 5]$, l'algoritmo restituisce "4" per la sottoseguenza crescente "1, 2, 3, 5".

**Comment by FeddyLix17** [comment]
È possibile implementare un algoritmo in $\Large O(n^2)$ utilizzando la tecnica della programmazione dinamica.

Tramite l'utilizzo di un'array ausiliario (memoization), **per ogni elemento dell'array A[]**, verrà tenuta traccia della lunghezza della massima sottoseguenza crescente possibile fino a quest'ulitmo.

Una volta calcolati i seguenti valori, verrà ritornato il più grande fra essi.

```python
def es1(A):

    # se l'array preso in input è vuoto
    if len(A) == 0:
        return 0

    # array per memorizzare le lunghezze delle massime sottoseguenze crescenti possibili fino a ciascun elemento
    L = [1] * len(A)

    # Calcolo della lunghezza della massima sottoseguenza crescente possibile per ciascun elemento.
    # Per ogni elemento dell'array A[] (tranne il primo)
    for i in range(1, len(A)):

        # effetuiamo un confronto con ogni elemento precedente ad esso
        for j in range(i):
            
            # e per ogni elemento minore riscontrato
            if A[i] > A[j]:
                
                # confrontiamo la lunghezza della massima sottoseguenza crescente possibile
                # tra quella dell'i-esimo elemento dell'array A[]
                # e quella dell'elemento minore riscontrato corrente
                # (se dovessimo aggiungere anche l'i-esimo elemento a quest'ultima)
                if L[i] < L[j] + 1:
                    # aggiornandola se neccessario
                    L[i] = L[j] + 1

    # ritorno il massimo valore dell'array L (anche se presente ripetutamente)
    maxLength = 0
    for x in L:
        if x > maxLength:
            maxLength = x

    return maxLength
```

Il costo dell'algoritmo è dato dalla somma dei costi di ogni blocco di codice

- il controllo della presenza di elementi nell'array A[] è elementare $\Large \Theta(1)$, mentre quello per costruire l'array ausiliario L $\Large \Theta(n)$,

    concludendo che il costo del primo blocco sia $\Large \Theta(1 + n) = \Theta(n)$.

- nei 2 cicli for annidati, verranno effettuate i operazioni elementari, per ogni i intero $\Large \in [1,\ n-1]$, arrivando a determinarne un costo

$$
    \Large \sum_{i\ =\ 1}^{n\ -\ 1} \Theta(i) = \Theta(\frac{(n-1)\ n}{2}) =  \Theta(\frac{n^2 - n}{2}) =  \Theta(n^2)
$$

- trovare il maggiore elemento presente nell'array L presenta costo lineare $\Large  \Theta(n)$.

Il costo finale dell'algoritmo risulta essere uguale a $\Large  \Theta(n) + \Theta(n^2) + \Theta(n) = \Theta(n^2)$.

Questo esercizio risulta essere un problema famoso denominato [Longest Increasing Subsequence](https://leetcode.com/problems/longest-increasing-subsequence/description/) (di cui esiste una versione più [avanzata](https://leetcode.com/problems/longest-increasing-subsequence-ii/description/)) il quale può essere [risolto](https://en.wikipedia.org/wiki/Longest_increasing_subsequence) con efficenza $\Large \Theta(n*log(n))$ utilizzando anche la ricerca binaria.

**Reply by Elia-Belli** [reply]
Una piccola precisazione, in $L[i]$ non è presente la lunghezza della massima sottosequenza crescente fino all'elemento $A[i]$, ma di una tale sottosequenza in cui $A[i]$ è compreso, quindi la soluzione finale non si trova in $L[n-1]$ ma si deve prendere $max(L)$: risulta poi evidente dal codice ma inizialmente mi aveva confuso 🙃

**Reply by cornflexxx** [reply]
Per ottenere il costo $O(n \log n)$ Possianmo pensare l'array $L[]$ in modo diverso:

$$L[i] = \text{valore minimo di $A$ in cui si ferma una sequenza di lunghezza $ i$}$$

In questo modo avremo che l'array $L$ sarà ordinato e per trovare la posizione in cui collocare $A[i]$ sarà sufficiente effettuare una ricerca binaria.
$L = [\infty,...\infty]$
$L[0] = -\infty$
Ad ogni iterazione troviamo l'indice in cui dovrebbe trovarsi $A[i]$, chiamiamo tale indice $q$, se 
$L[q-1] \lt A[i] \lt L[q]\Rightarrow L[q] = A[i]$ 
Dalla precedente costruzione di $L$ ne segue che la lunghezza della  sottosequenza massima è l'ultimo indice $i\lt len(L)|L[i]\neq \infty$.

**Reply by FeddyLix17** [reply]
@Elia-Belli alla fine dell'esecuzione dei vari cicli for l'array L[] conterrà al suo i-esimo elemento la massima sottosequenza crescente possibile tale che A[i] sia l'ultimo elemento di tale sottosequenza, e ciò NON implica che la massima sottosequenza crescente possibile di tutto l'array A[] abbia come ultimo elemento A[n -1]

**Comment by CiottoloMaggico** [comment]
![immagine](https://github.com/user-attachments/assets/ab41b806-a127-43f4-bcc9-e28e69546d1b)

**Comment by LorenzoBurello** [comment]
![es1](https://github.com/user-attachments/assets/db9bea4f-44e6-4073-bcd1-be1b3a3770ab)

**Comment by daromacarlo** [comment]
![immagine](https://github.com/user-attachments/assets/7084335b-82f8-4871-847f-6716a70417ee)

**Comment by niccolo-fato** [comment]
![photo_2025-06-23_21-17-20](https://github.com/user-attachments/assets/ab350e2f-5ac4-4490-873a-e44b2227e7db)
